editoast/src/views/train_schedule/projection.rs

```diff
@@ -234,17 +234,18 @@ async fn project_path(
             &path_routes,
             &path_blocks,
         );
-        trains_hash_values.insert(train.id, hash);
-        trains_details.insert(train.id, train_details);
+        trains_hash_values.push(hash);
+        trains_details.push(train_details);
     }
-    let cached_projections: Vec<Option<CachedProjectPathTrainResult>> = redis_conn
-        .json_get_bulk(&trains_hash_values.values().collect::<Vec<_>>())
-        .await?;
+    let cached_projections: Vec<Option<CachedProjectPathTrainResult>> =
+        redis_conn.json_get_bulk(&trains_hash_values).await?;
 
     let mut hit_cache: HashMap<i64, CachedProjectPathTrainResult> = HashMap::new();
     let mut miss_cache = HashMap::new();
-    for ((train_id, train_details), projection) in
-        trains_details.into_iter().zip(cached_projections)
+    for ((train_details, projection), train_id) in trains_details
+        .into_iter()
+        .zip(cached_projections)
+        .zip(trains.iter().map(|t| t.id))
```

Comment:
Just wondering if something like this is not more readable? (I know, it was already like this before the PR, just this PR is chaining it one more layer).

```rust
for (index, train_details) in trains_details.enumerate() {
  let train_id = trains[index].id;
  if let Some(cached) = cached_projections[index] {
    /* ... */
  }
  /* ... */
}
```

One drawback is that we used the `[]` indexing which can panic. But technically, it might be better than `.zip()` which will just silently short-circuit if one of the iterator is shorter.

Reply:
fixed